marconi-core/src/Marconi/Core/TracedStorable.hs

```diff
@@ -0,0 +1,391 @@
+module Marconi.Core.TracedStorable
+  ( -- * State
+    Config
+  , memoryBufferSize
+  , State
+  , handle
+  , config
+  , emptyState
+  , Storage
+  , storage
+  , events
+  , cursor
+  , getMemoryEvents
+  , getEvents
+  , StorableEvent
+  , StorablePoint
+  , StorableQuery
+  , StorableResult
+  , StorableMonad
+    -- * API
+  , SyntheticEvent(..)
+  , ControlNotification(..)
+  , Buffered(..)
+  , Queryable(..)
+  , Resumable(..)
+  , Rewindable(..)
+  , HasPoint(..)
+  , syntheticPoint
+  , foldEvents
+  , insert
+  , checkpoint
+  , insertMany
+  , rewind
+  , resume
+  , query
+  ) where
+
+import Control.Lens.Operators ((%~), (.~), (^.))
+import Control.Lens.TH qualified as Lens
+import Control.Monad (void)
+import Control.Monad.Primitive (PrimMonad, PrimState)
+import Data.Foldable (foldl', foldlM)
+import Data.Function ((&))
+import Data.Functor ((<&>))
+import Data.Vector qualified as V
+import Data.Vector.Generic qualified as VG
+import Data.Vector.Mutable qualified as VM
+import GHC.Generics (Generic)
+
+import Control.Tracer (Tracer, traceWith)
+
+{-
+   The extensible parts of the indexers are the way data is stored into some form
+   of persistent storage. The interface used for working blockchain events is defined
+   in this module.
+
+   There are quite a bit of type variables that are involed in defining this interface.
+   Some of them stand for the type of database connection (h), the type of events (e),
+   the type of points along the blockchain (denoted by p), the type of queries (q) and
+   results.
+
+   The following data families implement the observation that most of the variables
+   can be derived from the way information is stored in the database, so there is
+   quite a bit of convenience in reducing the number of type variables to just two.
+   One that stands for the database connection (which should be a newtype for each
+   indexer) and one for the monad in which the indexer runs (usually IO).
+-}
+data family StorableEvent h
+
+-- | The resume and query functionality requires a way to specify points on the chain from which we
+-- want to resume, or points up to which we want to query.
+type family StorablePoint h
+
+data family StorableQuery h
+
+data family StorableResult h
+
+-- TODO: Rename `Storable` to `Indexer`.
+-- data family StorableNotifications h
+
+data ControlNotification pt =
+    CNRollForward !pt
+  | CNRollBack    !pt
+ ---- | CNApplication !(StorableNotifications h)
+  deriving (Generic)
```

Comment:
Why is this one indexed with `pt` instead of `h`? Is it because of the type family for point ? My gut feeling is that adding the `CNAppplication` won't be trivial, I wonder if we should assess it now or postpone it. I think we'll have to add it somehow, but we may live without it in a near future.

Reply:
One thing I noticed is that using type applications inside a data makes life complicated. I prefer to use type variables and use the type application at the places where i use the data type.

This issue is somewhat incomplete. If we add notification, then we would need to add tracers to all the APIs defined by the type classes (so they can use the notification system as well). If I have the time to do it (and I probably will) I will do it next week.